File: anagrams-in-a-string.py
# Time Complexity: O(M+N) - M is len of s and N is len of p
# Space Complexity: O(N)
# Approach: Create a frequency counter for characters of the pattern. Use a window to iterate over the string. For every character, check if it matches the pattern, decrement counts in the frequency map to indicate exact matches. If the window length exceeds the length of the pattern, shrink the window by moving the start pointer and adding back the matched character frequencies to the map.
class Solution(object):
    def findAnagrams(self, s, p):
        """
        :type s: str
        :type p: str
        :rtype: List[int]
        """
        start = 0
        pattern_freq = collections.Counter(p)
        match = 0
        result = []
        
        for end in range(len(s)):
            end_char = s[end]
            if end_char in pattern_freq:
                pattern_freq[end_char] -= 1
                if pattern_freq[end_char] == 0:
                    # The characters in string in the window match pattern
                    match += 1
            # Window length exceeds pattern length, shrink window
            if end - start + 1 > len(p):
                start_char = s[start]
                start += 1
                if start_char in pattern_freq:
                    if pattern_freq[start_char] == 0:
                        match -= 1
                    pattern_freq[start_char] += 1
                    
            if len(pattern_freq) == match:
                result.append(start)
                
        return result

File: AnagramsInAString.java
import java.util.List;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Scanner;
public class AnagramsInAString {

        // SLIDING WINDOW ON STRINGS

        public List<Integer> findAnagrams(String s, String p) {

            List<Integer> result = new ArrayList<>();

            int m = p.length(); int n = s.length();

            // character frequency map of given pattern p
            HashMap<Character, Integer> freqmatch = new HashMap<>(); // O(1) space = 26 characters at maximum

            for(int i = 0; i < m; i++) { // O(m)

                char c = p.charAt(i);

                freqmatch.put(c, freqmatch.getOrDefault(c, 0) + 1);

            }

            // matching characters number is zero initially
            int match = 0;

            // iterate over all characters of string s
            for(int i = 0; i < n; i++) { // O(n)

                // incoming character in sliding window of pattern length
                char in = s.charAt(i);

                // if an incoming character is found in map, reduce its count
                if(freqmatch.containsKey(in)) {

                    int count = freqmatch.get(in);

                    count--;

                    // if count is zero, there is a match of incoming character in pattern
                    if(count == 0) {

                        match++;
                    }

                    //update map with new frequency of incoming character
                    freqmatch.put(in, count);
                }

                //outgoing
                //first outgoing character only when iterator i crosses pattern length
                if(i >= m) {

                    // (i - m) will be the index of outgoing character
                    char out = s.charAt(i - m);

                    // increase count of outgoing character if it is in map
                    if(freqmatch.containsKey(out)) {

                        int count = freqmatch.get(out);

                        count++;

                        // zero to one transition indicates match to mismatch transition
                        if(count == 1) {

                            match--;
                        }

                        //update map with new frequency of outgoing character
                        freqmatch.put(out, count);

                    }
                }

                // after every slide of window with iterator i moving forward, if match character count is same as map size of pattern -> anagram exists in that window
                if(match == freqmatch.size()) {

                    result.add(i - m + 1);
                }
            }
            //output list of starting indices of anagrams
            return result;
        }

        public static void main(String[] args) {

            AnagramsInAString object = new AnagramsInAString();

            Scanner scanner = new Scanner(System.in);

            // Prompt the user to enter string s
            System.out.print("Enter s: ");
            String s = scanner.nextLine();

            // Prompt the user to enter pattern p
            System.out.print("Enter p: ");
            String p = scanner.nextLine();

            List<Integer> answer = object.findAnagrams(s, p);

            System.out.println("Start indices of p's anagrams in s: ");

            for (Integer num : answer) {
                System.out.println(num);
            }
        }

}

/* TIME COMPLEXITY = O(M+N)
SPACE COMPLEXITY = O(1) */